pkg/ctl/create/utils.go

```diff
@@ -0,0 +1,230 @@
+package create
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/gobwas/glob"
+	"github.com/kris-nova/logger"
+	"github.com/pkg/errors"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/weaveworks/eksctl/pkg/ami"
+	api "github.com/weaveworks/eksctl/pkg/apis/eksctl.io/v1alpha4"
+	"github.com/weaveworks/eksctl/pkg/cfn/manager"
+	"github.com/weaveworks/eksctl/pkg/eks"
+)
+
+// NodeGroupFilter holds filter configuration
+type NodeGroupFilter struct {
+	IgnoreAllExisting bool
+
+	existing sets.String
+	only     []glob.Glob
+	onlySpec string
+}
+
+// NewNodeGroupFilter create new NodeGroupFilter instance
+func NewNodeGroupFilter() *NodeGroupFilter {
+	return &NodeGroupFilter{
+		IgnoreAllExisting: true,
+
+		existing: sets.NewString(),
+	}
+}
+
+// ApplyOnlyFilter parses given globs for exclusive filtering
+func (f *NodeGroupFilter) ApplyOnlyFilter(globs []string, cfg *api.ClusterConfig) error {
+	for _, gs := range globs {
+		g, err := glob.Compile(gs)
+		if err != nil {
+			return errors.Wrapf(err, "parsing glob filter %q", gs)
+		}
+		f.only = append(f.only, g)
+	}
+	f.onlySpec = strings.Join(globs, ",")
+	return f.onlyFilterMatchesAnything(cfg)
+}
+
+func (f *NodeGroupFilter) onlyFilterMatchesAnything(cfg *api.ClusterConfig) error {
+	if len(f.only) == 0 {
+		return nil
+	}
+	for i := range cfg.NodeGroups {
+		for _, g := range f.only {
+			if g.Match(cfg.NodeGroups[i].Name) {
+				return nil
+			}
+		}
+	}
+	return fmt.Errorf("no nodegroups match filter specification: %q", f.onlySpec)
+}
+
+// GetExisting uses stackManager to list existing nodegroup stacks and configures
+// the filter accordingly
+func (f *NodeGroupFilter) GetExisting(stackManager *manager.StackCollection) error {
+	if !f.IgnoreAllExisting {
+		return nil
+	}
+
+	existing, err := stackManager.ListNodeGroupStacks()
+	if err != nil {
+		return err
+	}
+
+	f.existing.Insert(existing...)
+
+	return nil
+}
+
+// Match checks given nodegroup against the filter
+func (f *NodeGroupFilter) Match(ng *api.NodeGroup) bool {
+	if f.IgnoreAllExisting && f.existing.Has(ng.Name) {
+		return false
+	}
+
+	for _, g := range f.only {
+		if g.Match(ng.Name) {
+			return true // return first match
+		}
+	}
+
+	// if no globs were given, match everything,
+	// if false - we haven't matched anything so far
+	return len(f.only) == 0
+}
+
+// MatchAll checks all nodegroups against the filter and returns all of
+// matching names as set
+func (f *NodeGroupFilter) MatchAll(cfg *api.ClusterConfig) sets.String {
+	names := sets.NewString()
+	for i := range cfg.NodeGroups {
+		if f.Match(cfg.NodeGroups[i]) {
+			names.Insert(cfg.NodeGroups[i].Name)
+		}
+	}
+	return names
+}
+
+// CheckEachNodeGroup iterates over each nodegroup and calls check function
+// (this is need to avoid common goroutine-for-loop pitfall)
+func CheckEachNodeGroup(f *NodeGroupFilter, cfg *api.ClusterConfig, check func(i int, ng *api.NodeGroup) error) error {
+	for i := range cfg.NodeGroups {
+		if f.Match(cfg.NodeGroups[i]) {
```

Comment:
Well, actually, this would look nicer if we do `ng := cfg.NodeGroups[i]` at the beginning, cause we now refer to it twice... I'll change it here, and above.

Reply:
Hm, why not just get it from ` range` like `for _, ng := range cfg.NodeGroups {...`

Reply:
That's to avoid [issue with callbacks](https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables).

Reply:
I cannot see any goroutine in here. What am I missing?

Reply:
You are right, there isn't one here. However, that issue applies when you have a closure, and I'm generally very precocious of it as I was bitten by it a number of times. I'll double check the history now, I think there was a closure before this code was moved around.

Reply:
I've added unit test and confirm this was indeed a hangover from the time when there used to be a closure.